src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -589,6 +592,40 @@ public void SendPendingBatches()
         _ = Renderer.Dispatcher.InvokeAsync(Renderer.ProcessBufferedRenderBatches);
     }
 
+    // Internal for testing.
+    internal Task DispatchInboundEventAsync(Func<Task> func)
+        => _dispatchInboundEvent(func);
+
+    // Internal for testing.
+    internal async Task<TResult> DispatchInboundEventAsync<TResult>(Func<Task<TResult>> func)
+    {
+        TResult result = default;
+        await _dispatchInboundEvent(async () => result = await func());
+        return result;
+    }
+
+    private static Func<Func<Task>, Task> BuildInboundEventDispatcher(IReadOnlyList<CircuitHandler> circuitHandlers, Circuit circuit)
+    {
+        Func<CircuitInboundEventContext, Task>? result = null;
+
+        for (var i = circuitHandlers.Count - 1; i >= 0; i--)
+        {
+            if (circuitHandlers[i] is IHandleCircuitEvent inboundEventHandler)
+            {
+                var next = result ?? (static (context) => context.Handler());
+                result = (context) => inboundEventHandler.HandleInboundEventAsync(context, next);
+            }
+        }
+
+        if (result is null)
+        {
+            // If there are no registered handlers, there is no need to allocate a context on each call.
+            return static (task) => task();
+        }
```

Comment:
This optimization was added because of the change making `CircuitInboundEventContext` a class.

Reply:
Ugh, just realized I should totally rename `task` to `func` here.